flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateFactory.java

```diff
@@ -0,0 +1,233 @@
+        ChangelogKeyGroupedPriorityQueue<T> changelogKeyGroupedPriorityQueue =
+                new ChangelogKeyGroupedPriorityQueue<>(internalPriorityQueue, logger, serializer);
+        priorityQueueStatesByName.put(stateName, changelogKeyGroupedPriorityQueue);
+        return changelogKeyGroupedPriorityQueue;
+    }
+
+    /**
+     * @param name state name
+     * @param type state type (the only supported type currently are: {@link
+     *     StateMetaInfoSnapshot.BackendStateType#KEY_VALUE key value}, {@link
+     *     StateMetaInfoSnapshot.BackendStateType#PRIORITY_QUEUE priority queue})
+     * @return an existing state, i.e. the one that was already created. The returned state will not
+     *     apply TTL to the passed values, regardless of the TTL settings. This prevents double
+     *     applying of TTL (recovered values are TTL values if TTL was enabled). The state will,
+     *     however, use TTL serializer if TTL is enabled. WARN: only valid during the recovery.
+     * @throws UnsupportedOperationException if state type is not supported
+     */
+    public ChangelogState getExistingState(String name, StateMetaInfoSnapshot.BackendStateType type)
+            throws UnsupportedOperationException {
+        ChangelogState state;
+        switch (type) {
+            case KEY_VALUE:
+                state = changelogStates.get(name);
+                break;
+            case PRIORITY_QUEUE:
+                state = priorityQueueStatesByName.get(name);
+                break;
+            default:
+                throw new UnsupportedOperationException(
+                        String.format("Unknown state type %s (%s)", type, name));
+        }
+        return state;
+    }
+
+    public void resetAllWritingMetaFlags() {
+        for (ChangelogState changelogState : changelogStates.values()) {
+            changelogState.resetWritingMetaFlag();
+        }
+
+        for (ChangelogKeyGroupedPriorityQueue<?> priorityQueueState :
+                priorityQueueStatesByName.values()) {
+            priorityQueueState.resetWritingMetaFlag();
+        }
+    }
+
+    public void dispose() {
+        changelogStates.clear();
+        priorityQueueStatesByName.clear();
+    }
+
+    private <S extends State, V> StateFactory getStateFactory(
+            StateDescriptor<S, V> stateDescriptor) {
+        StateFactory stateFactory = STATE_FACTORIES.get(stateDescriptor.getType());
+        if (stateFactory == null) {
+            String message =
+                    String.format(
+                            "State %s is not supported by %s",
+                            stateDescriptor.getClass(), ChangelogKeyedStateBackend.class);
+            throw new FlinkRuntimeException(message);
+        }
+        return stateFactory;
+    }
+
+    private static class VoidStateChangeLogger<Value, Namespace>
+            implements KvStateChangeLogger<Value, Namespace>, StateChangeLogger<Value, Namespace> {
```

Comment:
This logger is only used by `ChangelogMigrationRestoreTarget`, right?
If so, I'd move it there and inline the methods that are using it (`create`).